Force Unwrapping rule doesn't apply to self

### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Linting the following snippet with `force_unwrapping` enabled in `.swiftlint.yml`: 

```swift
class Test {
    func foo() {
        DispatchQueue.main.async { [weak self] in
            var string: String?
            _ = string!
            _ = self!
        }
    }
}
```

I would expect SwiftLint to warn on both force unwrappings. However, it does not catch `self!`.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint        
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths 
Linting 'foo.swift' (1/1)
/Users/simon/tmp/testswiftlint/foo.swift:5:23: warning: Force Unwrapping Violation: Force unwrapping should be avoided. (force_unwrapping)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

0.32.0 - installed with Homebrew

```yml
opt_in_rules:
  - force_unwrapping
```

Running Xcode 10.2.1 (confirmed with `xcode-select -p`), not using nested configurations.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -51,6 +51,11 @@
   [Sam Rayner](https://github.com/samrayner)
   [#2749](https://github.com/realm/SwiftLint/issues/2749)
 
+* Force-unwrapping `self` should trigger a violation of the `force_unwrapping`
+  rule.  
+  [Dalton Claybrook](https://github.com/daltonclaybrook)
+  [#2759](https://github.com/realm/SwiftLint/issues/2759)
+
 ## 0.32.0: Wash-N-Fold-N-Reduce
 
 #### Breaking

--- a/Rules.md
+++ b/Rules.md
@@ -8220,6 +8220,10 @@ func foo() -> [AnyHashable: Any]!
 func foo() -> [Int]! { return [] }
 ```
 
+```swift
+return self
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>
@@ -8276,6 +8280,10 @@ context("abc") {
 
 ```swift
 open var computed: String { return foo.bar↓! }
+```
+
+```swift
+return self↓!
 ```
 
 </details>

--- a/Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/ForceUnwrappingRule.swift
@@ -30,7 +30,8 @@ public struct ForceUnwrappingRule: OptInRule, ConfigurationProviderRule, Automat
             "func foo(_ options: [AnyHashable: Any]!) {",
             "func foo() -> [Int]!",
             "func foo() -> [AnyHashable: Any]!",
-            "func foo() -> [Int]! { return [] }"
+            "func foo() -> [Int]! { return [] }",
+            "return self"
         ],
         triggeringExamples: [
             "let url = NSURL(string: query)↓!",
@@ -49,7 +50,8 @@ public struct ForceUnwrappingRule: OptInRule, ConfigurationProviderRule, Automat
               var foo: [Bool]! = dict[\"abc\"]↓!
             }
             """,
-            "open var computed: String { return foo.bar↓! }"
+            "open var computed: String { return foo.bar↓! }",
+            "return self↓!"
         ]
     )
 
@@ -119,20 +121,18 @@ public struct ForceUnwrappingRule: OptInRule, ConfigurationProviderRule, Automat
                 .NSRangeToByteRange(start: secondRange.location, length: secondRange.length)
             else { return nil }
 
-        let kindsInFirstRange = syntaxMap.kinds(inByteRange: matchByteFirstRange)
-
         // check first captured range
-        // If not empty, first captured range is comment, string, keyword or typeidentifier.
+        // If not empty, first captured range is comment, string, typeidentifier or keyword that is not `self`.
         // We checks "not empty" because kinds may empty without filtering.
-        guard !kindsInFirstRange
-            .contains(where: ForceUnwrappingRule.excludingSyntaxKindsForFirstCapture.contains) else {
-                return nil
+        guard !isFirstRangeExcludedToken(byteRange: matchByteFirstRange, syntaxMap: syntaxMap, file: file) else {
+            return nil
         }
 
         let violationRange = NSRange(location: NSMaxRange(firstRange), length: 0)
+        let kindsInFirstRange = syntaxMap.kinds(inByteRange: matchByteFirstRange)
 
-        // if first captured range is identifier, generate violation
-        if kindsInFirstRange.contains(.identifier) {
+        // if first captured range is identifier or keyword (self), generate violation
+        if !Set(kindsInFirstRange).intersection([.identifier, .keyword]).isEmpty {
             return violationRange
         }
 
@@ -152,6 +152,22 @@ public struct ForceUnwrappingRule: OptInRule, ConfigurationProviderRule, Automat
         }
 
         return nil
+    }
+
+    // check if first captured range is comment, string, typeidentifier, or a keyword that is not `self`.
+    private func isFirstRangeExcludedToken(byteRange: NSRange, syntaxMap: SyntaxMap, file: File) -> Bool {
+        let tokens = syntaxMap.tokens(inByteRange: byteRange)
+        let nsString = file.contents.bridge()
+        return tokens.contains { token in
+            guard let kind = SyntaxKind(rawValue: token.type),
+                ForceUnwrappingRule.excludingSyntaxKindsForFirstCapture.contains(kind)
+                else { return false }
+            // check for `self
+            guard kind == .keyword,
+                let nsRange = nsString.byteRangeToNSRange(start: token.offset, length: token.length)
+                else { return true }
+            return nsString.substring(with: nsRange) != "self"
+        }
     }
 
     // check deepest kind matching range in structure is a typeAnnotation

